pkg/agent/util/iptables/iptables.go

```diff
@@ -167,64 +179,69 @@ func (c *Client) InsertRule(protocol Protocol, table string, chain string, ruleS
 		if err := ipt.Insert(table, chain, 1, ruleSpec...); err != nil {
 			return fmt.Errorf("error inserting rule %v to table %s chain %s: %v", ruleSpec, table, chain, err)
 		}
+		klog.V(2).Infof("Added rule %v to table %s chain %s of %s protocol", ruleSpec, table, chain, p.String())
+		return nil
 	}
-	klog.V(2).Infof("Inserted rule %v to table %s chain %s", ruleSpec, table, chain)
-	return nil
-}
-
-func matchProtocol(ipt *iptables.IPTables, protocol Protocol) bool {
-	switch protocol {
-	case ProtocolDual:
-		return true
-	case ProtocolIPv4:
-		return ipt.Proto() == iptables.ProtocolIPv4
-	case ProtocolIPv6:
-		return ipt.Proto() == iptables.ProtocolIPv6
+	if protocol == ProtocolDual {
+		if err := insertRule(ProtocolIPv4, table, chain, ruleSpec); err != nil {
+			return err
+		}
+		if err := insertRule(ProtocolIPv6, table, chain, ruleSpec); err != nil {
+			return err
+		}
+		return nil
 	}
-	return false
+	return insertRule(protocol, table, chain, ruleSpec)
 }
 
-// DeleteRule checks if target rule already exists, deletes the rule if found.
-func (c *Client) DeleteRule(table string, chain string, ruleSpec []string) error {
-	for idx := range c.ipts {
-		exist, err := c.ipts[idx].Exists(table, chain, ruleSpec...)
+// DeleteRules checks if target rule already exists in all iptables, deletes the rule if found.
+func (c *Client) DeleteRules(table string, chain string, ruleSpec []string) error {
+	for p := range c.ipts {
+		ipt := c.ipts[p]
+		exist, err := ipt.Exists(table, chain, ruleSpec...)
 		if err != nil {
 			return fmt.Errorf("error checking if rule %v exists in table %s chain %s: %v", ruleSpec, table, chain, err)
 		}
 		if !exist {
-			return nil
+			continue
 		}
-		if err := c.ipts[idx].Delete(table, chain, ruleSpec...); err != nil {
+		if err := ipt.Delete(table, chain, ruleSpec...); err != nil {
 			return fmt.Errorf("error deleting rule %v from table %s chain %s: %v", ruleSpec, table, chain, err)
 		}
-		klog.V(2).Infof("Deleted rule %v from table %s chain %s", ruleSpec, table, chain)
+		klog.V(2).Infof("Deleted rule %v from table %s chain %s of %s protocol", ruleSpec, table, chain, p.String())
 	}
 	return nil
 }
 
-// DeleteChain deletes all rules from a chain in a table and then delete the chain
-func (c *Client) DeleteChain(table string, chain string) error {
-	for idx := range c.ipts {
-		err := c.ipts[idx].ClearChain(table, chain)
+// DeleteChains deletes all rules from chains in all tables and then delete the chains.
+func (c *Client) DeleteChains(table string, chain string) error {
```

Comment:
Same comment as for `DeleteRules` (about the name)

Reply:
Updated to `DeleteMatchingChains`